catalog: start generifying desc.Collection for types

thoszhang opened the pull request:
This PR primarily extracts shared logic for accessing descriptors from `desc.Collection` previously used only for tables, to prepare for being able to access types. See individual commits for more details.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/51403)
<!-- Reviewable:end -->

rohany commented on pkg/sql/sqlbase/type_desc.go:
```diff
@@ -188,13 +208,33 @@ func (desc *MutableTypeDescriptor) MaybeIncrementVersion() {
 
+// OriginalName implements the MutableDescriptor interface.
+func (desc *MutableTypeDescriptor) OriginalName() string {
+	return desc.ClusterVersion.Name
```
The `clusterVersion` on a `MutableTypeDescriptor` is a pointer, so it'll be nil when the descriptor is new, so this might panic. Should you change it here to not be a pointer?

lucy-zhang replied:
Good point. The same is true for databases and schemas, and I think we should keep the `ClusterVersion` as pointers since these are immutable descriptors. I added nil checks for all of these fields.

In the future, though, I'd like to get rid of `ClusterVersion` since there's no good reason to keep the whole descriptor around. I think we really only need a few pieces of information (name, version, whether the descriptor is new, and maybe `NextMutationID`).

rohany commented on pkg/sql/table.go:
```diff
@@ -263,7 +263,7 @@ func (p *planner) writeTableDescToBatch(
 	}
 
-	if err := p.Tables().AddUncommittedTable(*tableDesc); err != nil {
+	if err := p.Tables().AddUncommittedDescriptor(tableDesc); err != nil {
```
can you add the corresponding call to add uncommitted descriptor in `writeTypeChange` in `type_change.go`

lucy-zhang replied:
Done.

rohany commented on pkg/sql/table.go:
```diff
@@ -263,7 +263,7 @@ func (p *planner) writeTableDescToBatch(
 	}
 
-	if err := p.Tables().AddUncommittedTable(*tableDesc); err != nil {
+	if err := p.Tables().AddUncommittedDescriptor(tableDesc); err != nil {
```
Also, maybe rename `Tables()` to `Descs()` here

lucy-zhang replied:
Done.

rohany reviewed (commented):
Nice! This looks like it'll be very useful to me as is. I'll just add some accessors on top that do the casting to `*TypeDescriptor` when I get there.

One thing I'm not sure about is how will we generalize this to schema and database descriptors. The `Get{Mutable}DescriptorVersion` functions that resolve based on name do extra filtering based on the qualified name, so we might have to add separate functions to get databases and schemas that are leased.

Commit:
catalog/descs: leasedTables -> leasedDescriptors

Release note: None

Commit:
sql: rename planner.Tables() to planner.Descriptors()

Release note: None

Commit:
catalog/descs: generify some Collection APIs and implementations

This commit extracts what is expected to be shared logic (either across
all descriptors, or for "objects," i.e., tables and types) from the
existing `Collection` methods, in preparation for supporting types (and,
after that, databases and schemas). In particular, it replaces
`uncommittedTables` with `uncommittedDescriptors` internally and
introduces `AddUncommittedDescriptor` and `HasUncommittedTypes`.

Release note: None

Commit:
sqlbase: add accessors for ClusterVersion state to MutableDescriptor

This commit adds accessors for some `ClusterVersion` fields to the
`MutableDescriptor` interface.

Release note: None

Commit:
sql: add modified TypeDescriptors to the descriptor collection

This commit starts adding modified `TypeDescriptor`s to the planner's
`desc.Collection`.

Release note: None

Commit:
catalogkv: avoid Validate() in GetMutableDescriptorByID

This commit is to deal with a bug. The current implementation of
`TableDescriptor.Validate()` always reads descriptors from the store
when validating cross-references. In the case where the descriptors
being read have been modified in the same transaction, a zero MVCC
timestamp is returned when reading the intent, so `ModificationTime`
can't be populated and a panic occurs. Currently it's particularly risky
to attempt to call `Validate()` when getting mutable descriptors as part
of a schema change. This commit replaces the call to `Validate()` with
`ValidateTable()`. The hope is that the table collection itself will be
responsible for both supplying descriptors to `Validate()` and
performing the validation.

Release note: None

Commit:
sqlbase: add descriptor add/drop/offline state accessors to interface

This commit lifts the `TableDescriptor`'s methods for descriptor state
to the `catalog.Descriptor` interface. It also makes
`sqlbase.FilterTableState`, which is used in table leasing, generic for
all descriptor types and moves it to `catalog`.

Release note: None

lucy-zhang commented:
> One thing I'm not sure about is how will we generalize this to schema and database descriptors. The Get{Mutable}DescriptorVersion functions that resolve based on name do extra filtering based on the qualified name, so we might have to add separate functions to get databases and schemas that are leased.

Yeah, that's what I was thinking for the methods that need to access descriptors by name.

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 19 files at r1, 4 of 5 files at r7, 1 of 1 files at r8, 6 of 11 files at r10, 22 of 22 files at r11, 1 of 1 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-:-MCmmsFk2cUv03muvL47:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang, @miretskiy, and @rohany)

___
*[pkg/sql/catalog/descriptor.go, line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-MCmjCZb3-QqLUxUcwZe:-MCmjCZb3-QqLUxUcwZf:b-o0zpa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c27cce91fa2da64be97192b9881531ebd4fff4f4/pkg/sql/catalog/descriptor.go#L29)):*
> ```
> // TODO (lucy): Make the error message specific to each descriptor type (e.g.,
> // "table is being dropped").
> var ErrDescriptorDropped = errors.New("descriptor is being dropped")
> ```
Also attach proper pgcodes codes. Can wait for later, maybe just update the TODO. Probably:

```
UndefinedDatabase                  = MakeCode("3D000")
	UndefinedSchema                    = MakeCode("3F000")
	UndefinedTable                     = MakeCode("42P01")
	UndefinedObject                    = MakeCode("42704")
```
or maybe for types:

```
	IndeterminateDatatype              = MakeCode("42P18")
```
___
*[pkg/sql/catalog/descs/collection.go, line 676 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-MCmlsgR3vNpcZ9JPu0z:-MCmlsgR3vNpcZ9JPu1-:bmkiaes) ([raw file](https://github.com/cockroachdb/cockroach/blob/c27cce91fa2da64be97192b9881531ebd4fff4f4/pkg/sql/catalog/descs/collection.go#L676)):*
> ```Go
> // transaction.
> // TODO (lucy): Figure out a more principled API for getting the tables updated
> // in the transaction, and maybe unify with GetTableDescsWithNewVersion.
> ```

I'm pro unification. Turning a slice of `ImmutableTableDescriptor` into a slice of `[]sqlbase.ID` is trivial. There's no way that extra allocation is a killer
___
*[pkg/sql/sqlbase/structured.go, line 3902 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-MCmmWsC8qO8yOHlJ6ut:-MCmmWsC8qO8yOHlJ6uu:bh98s97) ([raw file](https://github.com/cockroachdb/cockroach/blob/c27cce91fa2da64be97192b9881531ebd4fff4f4/pkg/sql/sqlbase/structured.go#L3902)):*
> ```Go
> 
> // Dropped returns whether the descriptor is dropped.
> // TODO (lucy): Does this method belong on Descriptor? This state does matter
> ```

I struggled with these too. This is pragmatic for now. 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/51403)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner, @miretskiy, and @rohany)

___
*[pkg/sql/catalog/descriptor.go, line 29 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-MCmjCZb3-QqLUxUcwZe:-MCnWDij1ZNU-19p7-EE:b-6xoeuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c27cce91fa2da64be97192b9881531ebd4fff4f4/pkg/sql/catalog/descriptor.go#L29)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> // TODO (lucy): Make the error message specific to each descriptor type (e.g.,
> // "table is being dropped").
> var ErrDescriptorDropped = errors.New("descriptor is being dropped")
> ```
Also attach proper pgcodes codes. Can wait for later, maybe just update the TODO. Probably:

```
UndefinedDatabase                  = MakeCode("3D000")
	UndefinedSchema                    = MakeCode("3F000")
	UndefinedTable                     = MakeCode("42P01")
	UndefinedObject                    = MakeCode("42704")
```
or maybe for types:

```
	IndeterminateDatatype              = MakeCode("42P18")
```
</blockquote></details>

Updated the TODO. It's not entirely clear to me whether `FilterDescriptorState` should be responsible for adding the pgcodes or whether its callers should just detect one of the errors returned and wrap the error, though I think the former is fine. Mostly, I don't actually like the idea of `FilterDescriptorState` having to generate a user-visible error and having to know the type of the descriptor, but it might be the most straightforward thing to do.
___
*[pkg/sql/catalog/descs/collection.go, line 676 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/51403#-MCmlsgR3vNpcZ9JPu0z:-MCmnsOP0bopPU-p2L2t:b1pwz48) ([raw file](https://github.com/cockroachdb/cockroach/blob/c27cce91fa2da64be97192b9881531ebd4fff4f4/pkg/sql/catalog/descs/collection.go#L676)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm pro unification. Turning a slice of `ImmutableTableDescriptor` into a slice of `[]sqlbase.ID` is trivial. There's no way that extra allocation is a killer
</blockquote></details>

Updated. I assume you mean unifying `GetTableIDsWithNewVersion` and `GetTableDescsWithNewVersion`? These methods return the complement of each other's table descriptor sets: the former returns only preexisting descriptors (for `validatePrimaryKeys`) and the latter returns only new descriptors (for `connExecutor.notifyStatsRefresherOfNewTables`).

Originally I was reluctant to combine these into a single method that returns all the uncommitted tables and have the caller figure out which tables are new, but looking at this again, it's fine. `validatePrimaryKeys` doesn't actually even need to only filter for existing tables; that was just an artifact of using preexisting methods, I think.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTRs

bors r+
